### PR TITLE
TDL-13997: Streams not honoring start date

### PR DESCRIPTION
**Pull request by harshpatel4crest**
# Description of change
TDL-13997: Streams not honoring start date
- Added code to save the tap state in a new state and not update the given state

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/test_harvest_new_start_date.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        cls._teardown_delete["clients"].append({"id": client_1["id"]})
+
+        project_1 = create_project(client_1['id'])
+        cls._teardown_delete["projects"].append({"id": project_1["id"]})
+
+        # Create dummy data in specifc streams     
+        for itter in range(3):
+            logging.info("Creating {} round(s) of data ...".format(itter + 1))
+
+            # estimate_1 = create_estimate(client_1['id'])
```
Comment (dbshah1212) [inline tests/test_harvest_new_start_date.py:39]: Can we remove commented code?

File: tests/test_harvest_new_start_date.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            self.assertGreater(record_count_sync, 1)
+
+            # check if all the invoices we created
+            # are collected in the invoice payment data
+            for all_invoice_id in all_invoice_ids:
+                self.assertTrue(all_invoice_id in invoice_ids_collected_from_sync)
+
+            second_updated_invoice_id = self.get_second_updated_at(sync_records)
+
+            # check if the second highest updated invoice is present
```
Comment (dbshah1212) [inline tests/test_harvest_new_start_date.py:99]: Instead of a second can we check all 3 invoices are available?

**Review by dbshah1212 — CHANGES_REQUESTED** [review]

File: tests/test_harvest_new_start_date.py (hunk 2 above)
Comment (harshpatel4crest) [reply tests/test_harvest_new_start_date.py:99]: Yes, It is checked at line #94

File: tests/test_harvest_new_start_date.py (hunk 1 above)
Comment (harshpatel4crest) [reply tests/test_harvest_new_start_date.py]: Done, added that stream

File: tap_harvest/__init__.py [hunk 3]
```diff
@@ -306,11 +308,11 @@ def map_invoice_payment(payment):
 
 
 def sync_estimates():
-    def map_estimate_message(message):
-        message['estimate_id'] = message['id']
-        return message
-
     def for_each_estimate(estimate, time_extracted):
+        def map_estimate_message(message):
```
Comment (dbshah1212) [inline tap_harvest/__init__.py:312]: Why we changed this code?

**Review by dbshah1212 — CHANGES_REQUESTED** [review]

File: tap_harvest/__init__.py (hunk 3 above)
Comment (harshpatel4crest) [reply tap_harvest/__init__.py:312]: Because in the records mapping was not right, `estimate_id` and `id` of each estimate message was the same

Comment (dbshah1212) [reply tap_harvest/__init__.py:312]: Can you put a small comment over it?

**Review by savan-chovatiya — APPROVED** [review]

File: tap_harvest/__init__.py [hunk 4]
```diff
@@ -23,6 +23,7 @@
 BASE_ID_URL = "https://id.getharvest.com/api/v2/"
 CONFIG = {}
 STATE = {}
+TAP_STATE = {}
```
Comment (karanpanchal-crest) [inline tap_harvest/__init__.py:26]: @harshpatel4crest Please add comments in the code

**Review by dbshah1212 — APPROVED** [review]

**Review by kpanchal-crest — APPROVED** [review]

**Review by karanpanchal-crest — APPROVED** [review]

**Review by KAllan357 — APPROVED** [review]

**Pull request merged**